platform/ios/src/MGLScaleBarView.m

```diff
@@ -0,0 +1,289 @@
+#import <Mapbox/Mapbox.h>
+
+#import "MGLScaleBarView_Private.h"
+
+@interface MGLScaleBarView()
+@property (nonatomic, assign, getter=isVisible) BOOL visible;
+@property (nonatomic, assign) CLLocationDistance metersPerPoint;
+@property (nonatomic) NSArray<UILabel *> *labels;
+@property (nonatomic) NSArray<UIView *> *bars;
+@property (nonatomic) UIView *containerView;
+@property (nonatomic) MGLDistanceFormatter *formatter;
+@property (nonatomic) NSArray<NSArray<NSNumber*>*> *metricTable;
+@property (nonatomic) NSArray<NSNumber*> *row;
+@property (nonatomic) UIColor *primaryColor;
+@property (nonatomic) UIColor *secondaryColor;
+@property (nonatomic) CALayer *borderLayer;
+@end
+
+static const CGFloat BAR_HEIGHT = 4;
+
+@implementation MGLScaleBarView
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    if (self = [super initWithCoder:aDecoder]) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    if (self = [super initWithFrame:frame]) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (void)commonInit {
+    self.clipsToBounds = NO;
+    self.alpha = 0;
+    
+    _visible = NO;
+    _primaryColor = [UIColor colorWithRed:62.0/255.0 green:62.0/255.0 blue:62.0/255.0 alpha:1];
+    _secondaryColor = [UIColor colorWithRed:247.0/255.0 green:247.0/255.0 blue:247.0/255.0 alpha:1];
```

Comment:
This will never be perfect, even if the developer gets full control over the scale bar’s colors. Consider Mapbox Satellite: a scale bar that’s readable over a lush forest may be unreadable over tundra and vice versa.

Apple Maps and Google Maps solve this problem by treating the scale bar’s labels just like any other labels on the base map: in Standard mode, the labels are set in black with a subtle, white stroke for contrast:

![ios-standard](https://cloud.githubusercontent.com/assets/1231218/21753865/8cf017bc-d5aa-11e6-8b81-4190bee050ab.PNG)
_Apple Maps on iOS, Standard_

<img width="247" alt="macos-standard" src="https://cloud.githubusercontent.com/assets/1231218/21753857/6fbe2094-d5aa-11e6-8330-28cafba66e7d.png"><br>
_Apple Maps on macOS, Standard_

![google-map](https://cloud.githubusercontent.com/assets/1231218/21753879/c0882af6-d5aa-11e6-8ecb-453b17b85397.PNG)
_Google Maps on iOS, main map_

![google-terrain](https://cloud.githubusercontent.com/assets/1231218/21753906/14ec226e-d5ab-11e6-87b9-e255222b1e4b.PNG)
_Google Maps on iOS, Terrain_

In Satellite mode, the labels are inverted to white with a subtle, black stroke. Another nice touch in Apple Maps is that the scale bar has slight translucency in Satellite mode to avoid becoming distracting.

![ios-satellite](https://cloud.githubusercontent.com/assets/1231218/21753921/455450b6-d5ab-11e6-9024-d8cef8d27aec.PNG)
_Apple Maps on iOS, Satellite_

<img width="246" alt="macos-satellite" src="https://cloud.githubusercontent.com/assets/1231218/21753929/56bd6e6e-d5ab-11e6-874d-c626b260e906.png"><br>
_Apple Maps on macOS, Satellite_

![google-satellite](https://cloud.githubusercontent.com/assets/1231218/21753911/2aca09de-d5ab-11e6-977f-1471fef9b547.PNG)
_Google Maps on iOS, Satellite_

Our SDK has to cope with a wider variety of color schemes and challenges, since the style could contain anything. Here are a few possibilities:

* Keep the appearance as-is. The scale bar isn’t the most important ornament on the screen; a little squinting may not be a showstopper.
* Place a blurred background beneath the scale bar, as the macOS SDK does for attribution. Note that this effect would be similar but not identical to a visual effect view, whose effect is a bit too vibrant for a control that shouldn’t be conspicuous.
* Choose between a light and dark appearance based on the style’s background color, as @friedbunny suggests.
* Find the most similar symbol style layer and mimic its appearance. This could yield the best results – the highest fidelity to the style – but could also be somewhat unpredictable.